### PR TITLE
[WASMFS] Mkdir Syscall

### PR DESCRIPTION
**Pull request by ethanalee**
Relevant Issue: #15041 

- Implement `mkdir`, `chdir`, cwd.
- TODO: Implement `unlink`, `rmdir`, `chmod` to complete `unlink.c` test.
- TODO: Extract functions where code is re-used.

### CODE REVIEW

File: system/lib/wasmfs/wasmfs.cpp [hunk 1]
```diff
@@ -303,4 +303,105 @@ __wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
...
+
+long __syscall_chdir(long path) {
+  std::vector<std::string> pathParts;
+
+  char newPathName[strlen((char*)path) + 1];
+  strcpy(newPathName, (char*)path);
+
+  char* current;
+
+  current = strtok(newPathName, "/\n");
```
Comment (ethanalee) [inline system/lib/wasmfs/wasmfs.cpp:369]: I think we can split out the path parsing and node traversal out now since it is used quite often. @azakai @tlively

File: system/lib/wasmfs/file_table.cpp [hunk 2]
```diff
@@ -117,6 +119,16 @@ std::shared_ptr<Directory> getRootDirectory() {
...
 
+std::shared_ptr<File> getCWD() {
+  if (cwd) {
+    return cwd;
+  } else {
+    return getRootDirectory();
+  }
+};
+
+void setCWD(std::shared_ptr<File> directory) { cwd = directory; };
```
Comment (kripken) [inline system/lib/wasmfs/file_table.cpp:130]: If these are the only two methods accessing `cwd` then it could be defined right next to them, which would make reading the code easier.

File: system/lib/wasmfs/file_table.h [hunk 3]
```diff
@@ -16,20 +16,49 @@
...
+  if (a > 0 && b > std::numeric_limits<T>::max() - a) {
+    return true;
+  }
+  return false;
+}
+
+std::shared_ptr<Directory> getParent(const std::vector<std::string>& pathParts,
+                                     long& err);
+
+std::vector<std::string> splitPath(char* pathname);
```
Comment (kripken) [inline system/lib/wasmfs/file_table.h:35]: Comments on such helpers would be good.

File: system/lib/wasmfs/file_table.h [hunk 4]
```diff
@@ -16,20 +16,49 @@
...
+
+// Access mode, file creation and file status flags for open.
+using wasmfs_oflags_t = uint32_t;
+
+extern std::shared_ptr<File> cwd;
```
Comment (kripken) [inline system/lib/wasmfs/file_table.h:40]: Why is this made accessible in this way?

File: system/lib/wasmfs/wasmfs.cpp [hunk 5]
```diff
@@ -61,31 +61,112 @@ __wasi_errno_t __wasi_fd_write(__wasi_fd_t fd,
...
+    if (result != __WASI_ERRNO_SUCCESS) {
+      *nread = offset - lockedOpenFile.position();
+      lockedOpenFile.position() = offset;
+      return result;
+    }
+    offset += bytesToRead;
+  }
+  *nread = offset - lockedOpenFile.position();
+  lockedOpenFile.position() = offset;
   return __WASI_ERRNO_SUCCESS;
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.cpp:170]: These are not related to mkdir, I think? Are they because of the difficulty in creating multiple open PRs? It does make it hard for me to know what to read carefully and what not. If we want multiple open PRs, could they target each other in a tree so the diffs are specific?

File: system/lib/wasmfs/wasmfs.cpp [hunk 6]
```diff
@@ -153,90 +234,138 @@ long __syscall_fstat64(long fd, long buf) {
...
+  auto openFile = std::make_shared<OpenFileState>(0, flags, curr);
+
+  return FileTable::get().add(openFile);
 }
 
-__wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
-  int accessMode = (flags & O_ACCMODE);
-  bool canWrite = false;
+long __syscall_mkdir(long path, long mode) {
+  std::vector<std::string> pathParts = splitPath((char*)path);
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.cpp:307]:
```suggestion
  auto pathParts = splitPath((char*)path);
```
I think these verbose types don't help readability, personally, in cases like this where it's clear we are returning a split path's parts.

File: system/lib/wasmfs/wasmfs.cpp [hunk 7]
```diff
@@ -153,90 +234,138 @@ long __syscall_fstat64(long fd, long buf) {
...
-  if (accessMode == O_WRONLY || accessMode == O_RDWR) {
-    canWrite = true;
+  // Root directory
+  if (pathParts.empty()) {
+    return -(EEXIST);
   }
 
-  std::vector<std::string> pathParts;
+  long err;
+  std::shared_ptr<Directory> parentDir = getParent(pathParts, err);
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.cpp:315]: How about renaming `getParent` to `getParentDir` and then using `auto` here? That would be shorter and clearer, and not ambiguous I think.

File: system/lib/wasmfs/wasmfs.cpp [hunk 8]
```diff
@@ -153,90 +234,138 @@ long __syscall_fstat64(long fd, long buf) {
...
-  char newPathName[strlen((char*)pathname) + 1];
-  strcpy(newPathName, (char*)pathname);
+  if (!parentDir) {
+    // parent node doesn't exist
+    return err;
+  }
+
+  auto lockedParentDir = parentDir->locked();
+
+  std::shared_ptr<File> curr =
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.cpp:324]: Same here with `auto`, I think that would be better.

File: system/lib/wasmfs/wasmfs.cpp [hunk 9]
```diff
@@ -153,90 +234,138 @@ long __syscall_fstat64(long fd, long buf) {
...
+    lockedParentDir.getEntry(pathParts[pathParts.size() - 1]);
 
-  // TODO: Support relative paths. i.e. specify cwd if path is relative.
-  // TODO: Other path parsing edge cases.
-  char* current;
+  // Check if the request directory already exists.
+  if (curr) {
+    return -(EEXIST);
+  } else {
+    // Create an empty in-memory file.
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.cpp:331]: This comment looks incorrect?

File: system/lib/wasmfs/wasmfs.cpp [hunk 10]
```diff
@@ -153,90 +234,138 @@ long __syscall_fstat64(long fd, long buf) {
...
+    auto created = std::make_shared<Directory>(mode);
 
-  current = strtok(newPathName, "/\n");
-  while (current != NULL) {
-    pathParts.push_back(current);
-    current = strtok(NULL, "/\n");
+    lockedParentDir.setEntry(pathParts[pathParts.size() - 1], created);
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.cpp:334]: `pathParts[pathParts.size() - 1]` appears twice and is not very clear by itself. How about using a temp var for it,
```
auto base = pathParts[pathParts.size() - 1];
```
(or "name", "basename", etc. - "base name" is what python calls it at least, not sure if there's not a better term).

File: system/lib/wasmfs/wasmfs.cpp [hunk 11]
```diff
@@ -153,90 +234,138 @@ long __syscall_fstat64(long fd, long buf) {
...
+    // parent node doesn't exist
+    return err;
+  }
 
-    // Requested entry (file or directory
-    if (!curr) {
-      return -(ENOENT);
-    }
+  std::shared_ptr<File> curr =
+    parentDir->locked().getEntry(pathParts[pathParts.size() - 1]);
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.cpp:357]: (auto please)

File: system/lib/wasmfs/file.cpp [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    buffer.resize(offset + len);
+    size = buffer.size();
+  }
+  memcpy(&buffer[offset], buf, len);
+
+  return __WASI_ERRNO_SUCCESS;
+}
+
+__wasi_errno_t MemoryFile::read(uint8_t* buf, size_t len, off_t offset) {
+  std::memcpy(buf, &buffer[offset], len);
```
[the remark this answers is not in the document]
Comment (ethanalee) [reply system/lib/wasmfs/file.cpp]: I think the caller already verifies this in `readAtOffset`: 

```
// Check if currOffset has exceeded size of file data.
    ssize_t dataLeft = size - currOffset;
    if (dataLeft <= 0) {
      break;
    }
```

Would an assert here be sufficient? Perhaps `assert(offset + len < buffer.size())`? @kripken

File: system/lib/wasmfs/wasmfs.cpp [hunk 13]
```diff
@@ -369,6 +355,72 @@ __wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
...
+    return err;
+  }
+
+  auto lockedParentDir = parentDir->locked();
+
+  auto curr = lockedParentDir.getEntry(base);
+
+  // Check if the request directory already exists.
+  if (curr) {
+    return -(EEXIST);
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.cpp:381]: Btw, why the `( )` here?

File: system/lib/wasmfs/wasmfs.cpp [hunk 14]
```diff
@@ -369,6 +355,72 @@ __wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
...
+    auto created = std::make_shared<Directory>(mode);
+
+    lockedParentDir.setEntry(base, created);
+    return 0;
+  }
+}
+
+long __syscall_chdir(long path) {
+  auto pathParts = splitPath((char*)path);
+  auto base = pathParts[pathParts.size() - 1];
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.cpp:393]: This line must appear after a check for `pathParts` being empty, or it will read invalid memory.

File: system/lib/wasmfs/wasmfs.cpp [hunk 15]
```diff
@@ -369,6 +355,72 @@ __wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
...
+
+  // Root directory
+  if (pathParts.empty()) {
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.cpp:396]: What about `chdir("/")`?

File: system/lib/wasmfs/wasmfs.cpp [hunk 16]
```diff
@@ -369,6 +355,72 @@ __wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
...
+    setCWD(getRootDirectory());
+    return 0;
+  }
+
+  long err;
+  auto parentDir = getParentDir(pathParts, err);
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.cpp:402]: Could we instead have a method `getDir` that would return a directory given a path? (or none if there is no such directory) Then this syscall and perhaps others wouldn't need to handle the parent at all.

File: system/lib/wasmfs/wasmfs.cpp (hunk 15 above)
Comment (kripken) [reply system/lib/wasmfs/wasmfs.cpp:396]: (The other code that checks for `"/"` suggests to me that this type of functionality should be shared.)

File: system/lib/wasmfs/wasmfs.cpp (hunk 13 above)
Comment (ethanalee) [reply system/lib/wasmfs/wasmfs.cpp:381]: I guess it is not really necessary. Thought it might have helped readability a little.
